numba/compiler.py

```diff
@@ -163,19 +162,67 @@ def __call__(self):
             res = self.func()
         except _EarlyEscape as e:
             return _StageResult(self, ok=True, result=e.result,
-                               early_escape=True)
+                                early_escape=True)
         except BaseException as e:
             return _StageResult(self, ok=False, exception=e)
         else:
             return _StageResult(self, ok=True, result=res)
 
 
-def _raise_error(desc, exc):
-    """Patches the error
-    """
-    newmsg = "{desc}\n{exc}".format(desc=desc, exc=exc)
-    exc.args = (newmsg,)
-    return exc
+class _PipelineManager(object):
+    def __init__(self):
+        self.pipeline_order = []
+        self.pipeline_stages = {}
+        self._finalized = False
+
+    def create_pipeline(self, pipeline_name):
+        assert not self._finalized, "Pipelines can no longer be added"
+        self.pipeline_order.append(pipeline_name)
+        self.pipeline_stages[pipeline_name] = []
+        self.current = pipeline_name
+
+    def add_stage(self, stage_function, stage_description):
+        assert not self._finalized, "Stages can no longer be added."
+        new_stage = _Stage(stage_function, stage_description)
+        current_pipeline_name = self.pipeline_order[-1]
+        self.pipeline_stages[current_pipeline_name].append(new_stage)
+
+    def finalize(self):
+        self._finalized = True
+
+    def _patch_error(self, desc, exc):
+        """
+        Patches the error
+        """
+        newmsg = "{desc}\n{exc}".format(desc=desc, exc=exc)
+        exc.args = (newmsg,)
+        return exc
+
+    def run(self, status):
+        assert self._finalized, "PM must be finalized before run()"
+        res = None
+        for pipeline_name in self.pipeline_order:
+            for stage in self.pipeline_stages[pipeline_name]:
+                res = stage()
+                # Stage failed?
+                if not res.ok:
+                    # No more fallback pipelines?
+                    msg = "Failed at " + stage.msg
+                    is_final_pipeline = pipeline_name == self.pipeline_order[-1]
+                    if is_final_pipeline:
+                        raise self._patch_error(msg, res.exception)
+                    # Go to next fallback pipeline
+                    else:
+                        status.fail_reason = self._patch_error(msg, res.exception)
+                        break
+                # Stage OK and early escape
+                elif res.early_escape:
+                    return res.result
+            else:
+                return res.result
+
+        # TODO save all error information
+        raise CompilerError("All pipelines have failed")
```

Comment:
Hmm.  Interesting.  Should I try to dissolve the exception-step in ```stage_objectmode_frontend()``` into a form more consistant with the pipeline design?  It seems funny to me to raise an exception that only gets caught in one specialized (already wrapped) function call as opposed to having a more communicative return type built into ```_Stage()``` or something similar.  

It is not completely clear to me if this is helpful, but at the very least, we can rename it something like "EarlyPipelineCompletion" which is a bit more descriptive of what it does.